diff --git a/app/CMakeLists.txt b/app/CMakeLists.txt
index 6a71e7f8..c7ac8c6b 100644
--- a/app/CMakeLists.txt
+++ b/app/CMakeLists.txt
@@ -98,3 +98,5 @@ target_sources(app PRIVATE src/main.c)
 add_subdirectory(src/display/)
 
 zephyr_cc_option(-Wfatal-errors)
+
+target_sources(app PRIVATE src/events/caps_word_state_changed.c)
diff --git a/app/include/zmk/events/caps_word_state_changed.h b/app/include/zmk/events/caps_word_state_changed.h
new file mode 100644
index 00000000..878f96b5
--- /dev/null
+++ b/app/include/zmk/events/caps_word_state_changed.h
@@ -0,0 +1,16 @@
+/*
+ * Copyright (c) 2020 The ZMK Contributors
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+#pragma once
+
+#include <zephyr/kernel.h>
+#include <zmk/event_manager.h>
+
+struct zmk_caps_word_state_changed {
+    bool state;
+};
+
+ZMK_EVENT_DECLARE(zmk_caps_word_state_changed);
diff --git a/app/src/behaviors/behavior_caps_word.c b/app/src/behaviors/behavior_caps_word.c
index d9b3f24e..a8d0855d 100644
--- a/app/src/behaviors/behavior_caps_word.c
+++ b/app/src/behaviors/behavior_caps_word.c
@@ -16,6 +16,7 @@
 #include <zmk/events/position_state_changed.h>
 #include <zmk/events/keycode_state_changed.h>
 #include <zmk/events/modifiers_state_changed.h>
+#include <zmk/events/caps_word_state_changed.h>
 #include <zmk/keys.h>
 #include <zmk/hid.h>
 #include <zmk/keymap.h>
@@ -41,16 +42,23 @@ struct behavior_caps_word_data {
     bool active;
 };
 
+int caps_word_raise_event(bool state) {
+    return raise_zmk_caps_word_state_changed(
+        (struct zmk_caps_word_state_changed){.state = state});
+}
+
 static void activate_caps_word(const struct device *dev) {
     struct behavior_caps_word_data *data = dev->data;
 
     data->active = true;
+    caps_word_raise_event(true);
 }
 
 static void deactivate_caps_word(const struct device *dev) {
     struct behavior_caps_word_data *data = dev->data;
 
     data->active = false;
+    caps_word_raise_event(false);
 }
 
 static int on_caps_word_binding_pressed(struct zmk_behavior_binding *binding,
diff --git a/app/src/events/caps_word_state_changed.c b/app/src/events/caps_word_state_changed.c
new file mode 100644
index 00000000..c2a83a19
--- /dev/null
+++ b/app/src/events/caps_word_state_changed.c
@@ -0,0 +1,4 @@
+#include <zephyr/kernel.h>
+#include <zmk/events/caps_word_state_changed.h>
+
+ZMK_EVENT_IMPL(zmk_caps_word_state_changed);
