diff --git a/app/src/activity.c b/app/src/activity.c
index 7894cb71..8a7dadc4 100644
--- a/app/src/activity.c
+++ b/app/src/activity.c
@@ -67,7 +67,7 @@ void activity_work_handler(struct k_work *work) {
     int32_t current = k_uptime_get();
     int32_t inactive_time = current - activity_last_uptime;
 #if IS_ENABLED(CONFIG_ZMK_SLEEP)
-    if (inactive_time > MAX_SLEEP_MS && !is_usb_power_present()) {
+    if (inactive_time > MAX_SLEEP_MS) {
         // Put devices in suspend power mode before sleeping
         set_state(ZMK_ACTIVITY_SLEEP);
         sys_poweroff();
