diff --git a/app/src/split/bluetooth/central.c b/app/src/split/bluetooth/central.c
index 54ac8131..8b81295f 100644
--- a/app/src/split/bluetooth/central.c
+++ b/app/src/split/bluetooth/central.c
@@ -28,6 +28,7 @@ LOG_MODULE_DECLARE(zmk, CONFIG_ZMK_LOG_LEVEL);
 #include <zmk/events/position_state_changed.h>
 #include <zmk/events/sensor_event.h>
 #include <zmk/events/battery_state_changed.h>
+#include <zmk/events/split_peripheral_status_changed.h>
 #include <zmk/hid_indicators_types.h>
 
 static int start_scanning(void);
@@ -144,6 +145,9 @@ int release_peripheral_slot(int index) {
     slot->update_hid_indicators = 0;
 #endif // IS_ENABLED(CONFIG_ZMK_SPLIT_PERIPHERAL_HID_INDICATORS)
 
+    raise_zmk_split_peripheral_status_changed(
+        (struct zmk_split_peripheral_status_changed){.connected = false});
+
     return 0;
 }
 
@@ -177,6 +181,9 @@ int confirm_peripheral_slot_conn(struct bt_conn *conn) {
     }
 
     peripherals[idx].state = PERIPHERAL_SLOT_STATE_CONNECTED;
+
+    raise_zmk_split_peripheral_status_changed(
+        (struct zmk_split_peripheral_status_changed){.connected = true});
     return 0;
 }
 
